src/ui/carousel/draggable-slide/slide-to-draggable-map.ts

```diff
@@ -0,0 +1,57 @@
+import { HorizontallyDraggable } from '../../draggable/horizontally-draggable';
+import { DefaultMap } from '../../../map/default-map';
+import { Draggable } from '../../draggable/draggable';
+import { Carousel } from '../carousel';
+import { Vector } from '../../../mathf/vector';
+import { mathf } from '../../..';
+import { Matrix } from './matrix';
+
+/**
+ * Given a carousel and draggable, constrain the given delta so that the
+ * draggable does not exceed the prescribed bounds of the carousel.
+ * @param carousel
+ * @param draggable
+ * @param delta
+ */
+function constrainDraggableSlide(
+    carousel: Carousel, draggable: Draggable, delta: Vector
+): Vector {
+  const slides = carousel.getSlides();
+
+  // Allow for centering the last slide
+  const halfContainer = carousel.getContainer().offsetWidth / 2;
+  const totalSlideWidth =
+      mathf.sum(slides.map((s) => s.offsetWidth));
+  const lastSlideWidth = slides.slice(-1)[0].offsetWidth;
+  const halfLastSlide = lastSlideWidth / 2;
+  const halfFirstSlide = slides[0].offsetWidth / 2;
+
+  const min = halfContainer - totalSlideWidth + halfLastSlide;
+  const max = halfContainer - halfFirstSlide;
+  const currentX =
+      Matrix.fromElementTransform(draggable.element).getTranslateX();
```

Comment:
Is there an easier way to get an element's X transform without having to create a Matrix()?

Reply:
Yes. We could parse only the X value from the result of `dom.getComputedStyle(el).transform`

The effort here was to create a canonical way of grabbing transform information so there would be a single point to update if browser quirks were introduced. However if we assume native functionality to be reliable I may be able to trim this Matrix content. I will investigate.

Reply:
Updated to use the MatrixService so it is grabbing the cached value.

So I re-read my code to remind myself why I had this complexity.

The MatrixService allows the dragging code to be separate from the slide/carousel specific code.

Without the MatrixService the code that loops slides asteroid-style overwrites the transformation applied by the Draggable code. It would also overwrite any styles on the element, though that may be less of a concern and any issue this creates could probably be resolved by just nesting another div.

So I could consolidate this all into a single DraggableSlide piece and remove the standalone Draggable and the need to have the Matrix/MatrixService if that's preferable, then deal only with the X component of the matrix and set a single inline style for `translateX()` rather than an inline style for the Matrix that preserves initial transforms set via CSS or other code acting on the element.

Reply:
Maybe dumb question but is this solved by adding say adding an inner div and adding one transform to the root and the other child?  If that is the case, I think I prefer adding the extra element over trying to solve matrix calculation conflicts to simply this...

(that being said, I think the matrix service class might be useful in cases where this is unavoidable so I think it would be nice to have that in degu eitherway)

Reply:
Yes, nesting the inner div solves this conflict.

The only advantage to nesting these divs would be it letting us remove the need for MatrixService, but if we want to keep MatrixService there's not as much benefit to streamlining to only `translateX()` styles.